google/cloud/pubsub/internal/tracing_batch_callback.cc

```diff
@@ -0,0 +1,146 @@
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+namespace {
+
+opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> StartSubscribeSpan(
+    google::pubsub::v1::ReceivedMessage const& message,
+    pubsub::Subscription const& subscription,
+    std::shared_ptr<
+        opentelemetry::context::propagation::TextMapPropagator> const&
+        propagator) {
+  opentelemetry::trace::StartSpanOptions options;
+  options.kind = opentelemetry::trace::SpanKind::kConsumer;
+  auto m = pubsub_internal::FromProto(message.message());
+  auto context = ExtractTraceContext(m, *propagator);
+  auto producer_span_context =
+      opentelemetry::trace::GetSpan(context)->GetContext();
+  if (producer_span_context.IsSampled() && producer_span_context.IsValid()) {
+    options.parent = producer_span_context;
+  }
+
+  auto span = internal::MakeSpan(subscription.subscription_id() + " subscribe",
+                                 options);
+  return span;
+}
+
+}  // namespace
+
+/**
+ * Tracing batch callback implementation.
+ * */
+class TracingBatchCallback : public BatchCallback {
+ public:
+  explicit TracingBatchCallback(std::shared_ptr<BatchCallback> child,
+                                pubsub::Subscription const& subscription)
+      : child_(std::move(child)),
+        subscription_(std::move(subscription)),
+        propagator_(std::make_shared<
+                    opentelemetry::trace::propagation::HttpTraceContext>()) {}
+  ~TracingBatchCallback() override {
+    std::lock_guard<std::mutex> lk(mu_);
+    // End all outstanding subscribe spans.
+    for (auto const& kv : subscribe_span_by_ack_id_) {
+      kv.second->End();
+    }
+    subscribe_span_by_ack_id_.clear();
+  }
+
+  void callback(BatchCallback::StreamingPullResponse response) override {
+    if (response.response) {
+      for (auto const& message : response.response->received_messages()) {
+        auto subscribe_span =
+            StartSubscribeSpan(message, subscription_, propagator_);
+        auto scope = internal::OTelScope(subscribe_span);
+        {
+          std::lock_guard<std::mutex> lk(mu_);
+          subscribe_span_by_ack_id_[message.ack_id()] = subscribe_span;
+        }
+      }
+    }
+    child_->callback(std::move(response));
+  };
+
+  void message_callback(MessageCallback::ReceivedMessage m) override {
+    child_->message_callback(std::move(m));
+  }
+
+  void user_callback(MessageCallback::MessageAndHandler m) override {
+    child_->user_callback(std::move(m));
+  }
+
+  void EndMessage(std::string const& ack_id,
+                  std::string const& event) override {
+    std::lock_guard<std::mutex> lk(mu_);
+    // Use the ack_id to find the subscribe span and end it.
+    auto subscribe_span = subscribe_span_by_ack_id_.find(ack_id);
+    if (subscribe_span != subscribe_span_by_ack_id_.end()) {
+      subscribe_span->second->AddEvent(event);
+      subscribe_span->second->End();
+      subscribe_span_by_ack_id_.erase(ack_id);
```

Comment:
Erase by iterator to avoid a second lookup.

Reply:
Done